# Problem 6

In [47]:
# make pandas dataframe of information
import numpy as np
import pandas as pd
from icecream import ic

# ,columns=["f[xi,xi+1]","f[xi,xi+1,xi+2]"]

df = pd.DataFrame({"x_i" : [-0.1,0.0,0.2,0.3], "f(xi)" : [5.3,2.0,3.19,1.0], "f[xi-1,xi]": np.nan , "f[xi-2,xi-1,xi]" : np.nan, "f[xi-3,xi-2,xi-1,xi]" : np.nan})

# make first col of differences
for i in range(1,4):
    # make the row in col 2, f[xi-1,xi] = (f(xi) - f(xi-1)) / (xi - xi-1)
    df.iloc[i,2] = (df.iloc[i, 1] - df.iloc[i-1,1]) / (df.iloc[i,0] - df.iloc[i-1,0])

# make second col of differences
for i in range(2,4):
    # make the row in colindx 3, f[xi-1,xi] = (f(xi) - f(xi-1)) / (xi - xi-1)
    df.iloc[i,3] = (df.iloc[i, 2] - df.iloc[i-1,2]) / (df.iloc[i,0] - df.iloc[i-2,0])

# 3rd col
for i in range(3,4):
    df.iloc[i,4] = (df.iloc[i, 3] - df.iloc[i-1,3]) / (df.iloc[i,0] - df.iloc[i-3,0])

df

x_i  f(xi)  f[xi-1,xi]  f[xi-2,xi-1,xi]  f[xi-3,xi-2,xi-1,xi]
0 -0.1   5.30         NaN              NaN                   NaN
1  0.0   2.00      -33.00              NaN                   NaN
2  0.2   3.19        5.95       129.833333                   NaN
3  0.3   1.00      -21.90       -92.833333           -556.666667

Based on the produced chart of differences, the following will be the interpolating polynomial of the given points:

\begin{gather*}

P_3(x) = a_0 + a_1 (x-x_0) + a_2 (x-x_0)(x-x_1) + a_3 (x-x_0)(x-x_1)(x-x_2) \\ \implies \\

P_3(x) = f[x_0] + f[x_0,x_1] (x-x_0) + f[x_0,x_1,x_2] (x-x_0)(x-x_1) + f[x_0,x_1,x_2,x_3] (x-x_0)(x-x_1)(x-x_2) \\ \implies \\

P_3(x) = 5.3 + -33 (x+0.1) + 129.833 (x+0.1)(x) -556.667 (x+0.1)(x)(x-0.2) \\ \implies \\

\end{gather*}